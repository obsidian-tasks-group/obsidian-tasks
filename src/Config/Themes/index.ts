export * from './AnuPpuccinThemeCollection';
export * from './AuraThemeCollection';
export * from './EbullientworksThemeCollection';
export * from './ITSThemeCollection';
export * from './LYTModeThemeCollection';
export * from './MinimalThemeCollection';
export * from './ThingsThemeCollection';
