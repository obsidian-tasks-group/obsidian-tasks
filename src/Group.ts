import { GroupHeadings, IntermediateTaskGroups } from './GroupDetails';
import type { Grouping, GroupingProperty } from './Query';
import type { Task } from './Task';

/**
 * A naming function, that takes a Task object and returns the corresponding group property name
 */
type Grouper = (task: Task) => string;

/**
 * Implementation of the 'group by' instruction.
 */
export class Group {
    /**
     * Group a list of tasks, according to one or more task properties
     * @param grouping 0 or more Grouping values, one per 'group by' line
     * @param tasks The tasks that match the task block's Query
     */
    public static by(grouping: Grouping[], tasks: Task[]): TaskGroups {
        return new TaskGroups(grouping, tasks);
    }

    /**
     * Return the Grouper functions matching the 'group by' lines
     * @param grouping 0 or more Grouping values, one per 'group by' line
     */
    public static getGroupersForAllQueryGroupings(grouping: Grouping[]) {
        const groupers: Grouper[] = [];
        for (const { property } of grouping) {
            const comparator = Group.groupers[property];
            groupers.push(comparator);
        }
        return groupers;
    }

    /**
     * Return the group names for a single task
     * @param groupers The Grouper functions indicating the requested types of group
     * @param task
     */
    public static getGroupNamesForTask(groupers: Grouper[], task: Task) {
        const groupNames = [];
        for (const grouper of groupers) {
            const groupName = grouper(task);
            groupNames.push(groupName);
        }
        return groupNames;
    }

    /**
     * Return a single property name for a single task.
     * A convenience method for unit tests.
     * @param property
     * @param task
     */
    public static getGroupNameForTask(
        property: GroupingProperty,
        task: Task,
    ): string {
        const grouper = Group.groupers[property];
        return grouper(task);
    }

    private static groupers: Record<GroupingProperty, Grouper> = {
        backlink: Group.groupByBacklink,
        filename: Group.groupByFileName,
        folder: Group.groupByFolder,
        heading: Group.groupByHeading,
        path: Group.groupByPath,
        status: Group.groupByStatus,
    };

    private static groupByPath(task: Task): string {
        // Does this need to be made stricter?
        // Is there a better way of getting the file name?
        return task.path.replace('.md', '');
    }

    private static groupByFolder(task: Task): string {
        const path = task.path;
        const fileNameWithExtension = task.filename + '.md';
        const folder = path.substring(
            0,
            path.lastIndexOf(fileNameWithExtension),
        );
        if (folder === '') {
            return '/';
        }
        return folder;
    }

    private static groupByFileName(task: Task): string {
        // Note current limitation: Tasks from different notes with the
        // same name will be grouped together, even though they are in
        // different files and their links will look different.
        const filename = task.filename;
        if (filename == null) {
            return 'Unknown Location';
        }
        return filename;
    }

    private static groupByBacklink(task: Task): string {
        const linkText = task.getLinkText({ isFilenameUnique: true });
        if (linkText == null) {
            return 'Unknown Location';
        }
        return linkText;
    }

    private static groupByStatus(task: Task): string {
        return task.status;
    }

    private static groupByHeading(task: Task): string {
        if (task.precedingHeader == null || task.precedingHeader.length == 0) {
            return '(No heading)';
        }
        return task.precedingHeader;
    }
}

/**
 * GroupHeading contains the data needed to render one heading for a group of tasks
 */
export class GroupHeading {
    /**
     * How nested the heading is.
     * 0 is the first group, meaning this heading was generated by
     * the first 'group by' instruction.
     */
    public readonly nestingLevel: number;

    /**
     * The text to be displayed for the group.
     */
    public readonly name: string;

    /**
     * Construct a GroupHeading object
     * @param {number} nestingLevel - See this.nestingLevel for details
     * @param {string} name - The text to be displayed for the group
     */
    constructor(nestingLevel: number, name: string) {
        this.nestingLevel = nestingLevel;
        this.name = name;
    }
}

/**
 * TaskGroup stores a single group of tasks, that all share the same group names.
 * TaskGroup objects are stored in a TaskGroups object.
 *
 * For example, if the user supplied these 'group by' lines:
 *   group by folder
 *   group by filename
 *   group by heading
 * Then the names of one TaskGroup might be this:
 *   Some/Folder/In/The/Vault
 *   A Particular File Name
 *   My lovely heading
 * And the TaskGroup would store all the tasks from that location
 * that match the task block's filters, in the task block's sort order
 */
export class TaskGroup {
    /**
     * The names of the group properties for this set of tasks,
     * in the order of the 'group by' lines the user specified
     */
    public readonly groups: string[];

    /**
     * The headings to be displayed in front of this set of tasks,
     * when rendering the results.
     *
     * It only contains the minimal set of headings required to separate
     * this group of tasks from the previous group of tasks.
     */
    public readonly groupHeadings: GroupHeading[];

    /**
     * All the tasks that match the user's filters and that have the
     * group names exactly matching groups().
     */
    public readonly tasks: Task[];

    /**
     * Constructor for TaskGroup
     * @param {string[]} groups - See this.groups for details
     * @param {GroupHeading[]} groupHeadings - See this.groupHeadings for details
     * @param tasks {Task[]} - See this.tasks for details
     */
    constructor(
        groups: string[],
        groupHeadings: GroupHeading[],
        tasks: Task[],
    ) {
        this.groups = groups;
        this.groupHeadings = groupHeadings;
        this.tasks = tasks;
    }

    /**
     * A markdown-format representation of all the tasks in this group.
     *
     * Useful for testing.
     */
    public tasksAsStringOfLines(): string {
        let output = '';
        for (const task of this.tasks) {
            output += task.toFileLineString() + '\n';
        }
        return output;
    }

    /**
     * A human-readable representation of this task group, including names
     * and headings that should be displayed.
     *
     * Note that this is used in snapshot testing, so if the format is
     * changed, the snapshots will need to be updated.
     */
    public toString(): string {
        let output = '\n';
        output += `Group names: [${this.groups}]\n`;

        for (const heading of this.groupHeadings) {
            // These headings mimic the behaviour of QueryRenderer,
            // which uses 'h4', 'h5' and 'h6' for nested groups.
            const headingPrefix = '#'.repeat(4 + heading.nestingLevel);
            output += `${headingPrefix} ${heading.name}\n`;
        }

        output += this.tasksAsStringOfLines();
        return output;
    }
}

/**
 * TaskGroup stores all the groups of tasks generated by any 'group by'
 * instructions in the task block.
 */
export class TaskGroups {
    private _groups: TaskGroup[] = new Array<TaskGroup>();

    /**
     * Constructor for TaskGroups
     * @param {Grouping[]} groups - 0 or more Grouping values,
     *                              1 per 'group by' line in the task query block
     * @param {Task[]} tasks - 0 more more Task objects, with all the tasks
     *                         matching the query, already in sort order
     */
    constructor(groups: Grouping[], tasks: Task[]) {
        const initialGroups = new IntermediateTaskGroups(groups, tasks);
        this.addTasks(initialGroups);
    }

    /**
     * All the tasks matching the query, grouped together, and in the order
     * that they should be displayed.
     */
    public get groups(): TaskGroup[] {
        return this._groups;
    }

    /**
     * The total number of tasks matching the query.
     */
    public totalTasksCount() {
        let totalTasksCount = 0;
        for (const group of this.groups) {
            totalTasksCount += group.tasks.length;
        }
        return totalTasksCount;
    }

    /**
     * A human-readable representation of all the task groups.
     *
     * Note that this is used in snapshot testing, so if the format is
     * changed, the snapshots will need to be updated.
     */
    public toString(): string {
        let output = '';
        for (const taskGroup of this.groups) {
            output += taskGroup.toString();
            output += '\n---\n';
        }
        const totalTasksCount = this.totalTasksCount();
        output += `\n${totalTasksCount} tasks\n`;
        return output;
    }

    private addTasks(initialGroups: IntermediateTaskGroups) {
        // Get the headings
        const grouper = new GroupHeadings(initialGroups.groups);

        // Build a container of all the groups
        for (const [groups, tasks] of initialGroups.groups) {
            const groupHeadings = grouper.getHeadingsForTaskGroup(groups);
            const taskGroup = new TaskGroup(groups, groupHeadings, tasks);
            this.add(taskGroup);
        }
    }

    private add(taskGroup: TaskGroup) {
        this._groups.push(taskGroup);
    }
}
