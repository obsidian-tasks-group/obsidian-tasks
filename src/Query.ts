import chrono from 'chrono-node';
import { LayoutOptions } from './LayoutOptions';

import { Status, Task } from './Task';

type Sorting = 'status' | 'due' | 'done' | 'path' | 'description';

export class Query {
    private _limit: number | undefined = undefined;
    private _layoutOptions: LayoutOptions = new LayoutOptions();
    private _filters: ((task: Task) => boolean)[] = [];
    private _error: string | undefined = undefined;
    private _sorting: Sorting[] = [];

    private readonly noDueString = 'no due date';
    private readonly dueRegexp = /^due (before|after|on)? ?(.*)/;

    private readonly doneString = 'done';
    private readonly notDoneString = 'not done';
    private readonly doneRegexp = /^done (before|after|on)? ?(.*)/;

    private readonly pathRegexp = /^path (includes|does not include) (.*)/;
    private readonly descriptionRegexp =
        /^description (includes|does not include) (.*)/;
    private readonly sortByRegexp =
        /^sort by (status|due|done|path|description)/;

    private readonly headingRegexp =
        /^heading (includes|does not include) (.*)/;

    private readonly hideOptionsRegexp =
        /^hide (task count|backlink|done date|due date|recurrence rule|edit button)/;

    private readonly recurringString = 'is recurring';
    private readonly notRecurringString = 'is not recurring';

    private readonly limitRegexp = /^limit (to )?(\d+)( tasks?)?/;
    private readonly excludeSubItemsString = 'exclude sub-items';

    constructor({ source }: { source: string }) {
        source
            .split('\n')
            .map((line: string) => line.trim())
            .forEach((line: string) => {
                switch (true) {
                    case line === '':
                        break;
                    case line === this.doneString:
                        this._filters.push(
                            (task) => task.status === Status.Done,
                        );
                        break;
                    case line === this.notDoneString:
                        this._filters.push(
                            (task) => task.status !== Status.Done,
                        );
                        break;
                    case line === this.recurringString:
                        this._filters.push(
                            (task) => task.recurrenceRule !== null,
                        );
                        break;
                    case line === this.notRecurringString:
                        this._filters.push(
                            (task) => task.recurrenceRule === null,
                        );
                        break;
                    case line === this.excludeSubItemsString:
                        this._filters.push((task) => task.indentation === '');
                        break;
                    case line === this.noDueString:
                        this._filters.push((task) => task.dueDate === null);
                        break;
                    case this.dueRegexp.test(line):
                        this.parseDueFilter({ line });
                        break;
                    case this.doneRegexp.test(line):
                        this.parseDoneFilter({ line });
                        break;
                    case this.pathRegexp.test(line):
                        this.parsePathFilter({ line });
                        break;
                    case this.descriptionRegexp.test(line):
                        this.parseDescriptionFilter({ line });
                        break;
                    case this.headingRegexp.test(line):
                        this.parseHeadingFilter({ line });
                        break;
                    case this.limitRegexp.test(line):
                        this.parseLimit({ line });
                        break;
                    case this.sortByRegexp.test(line):
                        this.parseSortBy({ line });
                        break;
                    case this.hideOptionsRegexp.test(line):
                        this.parseHideOptions({ line });
                        break;
                    default:
                        this._error = 'do not understand query';
                }
            });
    }

    public get limit(): number | undefined {
        return this._limit;
    }

    public get layoutOptions(): LayoutOptions {
        return this._layoutOptions;
    }

    public get filters(): ((task: Task) => boolean)[] {
        return this._filters;
    }

    public get sorting(): Sorting[] {
        return this._sorting;
    }

    public get error(): string | undefined {
        return this._error;
    }

    private parseHideOptions({ line }: { line: string }): void {
        const hideOptionsMatch = line.match(this.hideOptionsRegexp);
        if (hideOptionsMatch !== null) {
            const option = hideOptionsMatch[1].trim().toLowerCase();

            if (option === 'task count') {
                this._layoutOptions.hideTaskCount = true;
            } else if (option === 'backlink') {
                this._layoutOptions.hideBacklinks = true;
            } else if (option === 'done date') {
                this._layoutOptions.hideDoneDate = true;
            } else if (option === 'due date') {
                this._layoutOptions.hideDueDate = true;
            } else if (option === 'recurrence rule') {
                this._layoutOptions.hideRecurrenceRule = true;
            } else if (option === 'edit button') {
                this._layoutOptions.hideEditButton = true;
            } else {
                this._error = 'do not understand hide option';
            }
        }
    }

    private parseDueFilter({ line }: { line: string }): void {
        const dueMatch = line.match(this.dueRegexp);
        if (dueMatch !== null) {
            const filterDate = this.parseDate(dueMatch[2]);
            if (!filterDate.isValid()) {
                this._error = 'do not understand due date';
            }

            let filter;
            if (dueMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isBefore(filterDate) : false;
            } else if (dueMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isSame(filterDate) : false;
            }

            this._filters.push(filter);
        } else {
            this._error = 'do not understand query filter (due date)';
        }
    }

    private parseDoneFilter({ line }: { line: string }): void {
        const doneMatch = line.match(this.doneRegexp);
        if (doneMatch !== null) {
            const filterDate = this.parseDate(doneMatch[2]);
            if (!filterDate.isValid()) {
                this._error = 'do not understand done date';
            }

            let filter;
            if (doneMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isBefore(filterDate) : false;
            } else if (doneMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isSame(filterDate) : false;
            }

            this._filters.push(filter);
        }
    }

    private parsePathFilter({ line }: { line: string }): void {
        const pathMatch = line.match(this.pathRegexp);
        if (pathMatch !== null) {
            const filterMethod = pathMatch[1];
            if (filterMethod === 'includes') {
                this._filters.push((task: Task) =>
                    task.path.includes(pathMatch[2]),
                );
            } else if (pathMatch[1] === 'does not include') {
                this._filters.push(
                    (task: Task) => !task.path.includes(pathMatch[2]),
                );
            } else {
                this._error = 'do not understand query filter (path)';
            }
        } else {
            this._error = 'do not understand query filter (path)';
        }
    }

    private parseDescriptionFilter({ line }: { line: string }): void {
        const descriptionMatch = line.match(this.descriptionRegexp);
        if (descriptionMatch !== null) {
            const filterMethod = descriptionMatch[1];
            if (filterMethod === 'includes') {
                this._filters.push((task: Task) =>
                    this.stringIncludesCaseInsensitive(
                        task.description,
                        descriptionMatch[2],
                    ),
                );
            } else if (descriptionMatch[1] === 'does not include') {
                this._filters.push(
                    (task: Task) =>
                        !this.stringIncludesCaseInsensitive(
                            task.description,
                            descriptionMatch[2],
                        ),
                );
            } else {
                this._error = 'do not understand query filter (description)';
            }
        } else {
            this._error = 'do not understand query filter (description)';
        }
    }

    private parseHeadingFilter({ line }: { line: string }): void {
        const headingMatch = line.match(this.headingRegexp);
        if (headingMatch !== null) {
            const filterMethod = headingMatch[1].toLowerCase();
            if (filterMethod === 'includes') {
                this._filters.push(
                    (task: Task) =>
                        task.precedingHeader !== null &&
                        this.stringIncludesCaseInsensitive(
                            task.precedingHeader,
                            headingMatch[2],
                        ),
                );
            } else if (headingMatch[1] === 'does not include') {
                this._filters.push(
                    (task: Task) =>
                        task.precedingHeader !== null &&
                        !this.stringIncludesCaseInsensitive(
                            task.precedingHeader,
                            headingMatch[2],
                        ),
                );
            } else {
                this._error = 'do not understand query filter (heading)';
            }
        } else {
            this._error = 'do not understand query filter (heading)';
        }
    }

    private parseLimit({ line }: { line: string }): void {
        const limitMatch = line.match(this.limitRegexp);
        if (limitMatch !== null) {
            // limitMatch[2] is per regex always digits and therefore parsable.
            const limit = Number.parseInt(limitMatch[2], 10);
            this._limit = limit;
        } else {
            this._error = 'do not understand query limit';
        }
    }

    private parseSortBy({ line }: { line: string }): void {
        const fieldMatch = line.match(this.sortByRegexp);
        if (fieldMatch !== null) {
            this._sorting.push(fieldMatch[1] as Sorting);
        } else {
            this._error = 'do not understand query sorting';
        }
    }

    private parseDate(input: string): moment.Moment {
        // Using start of date to correctly match on comparison with other dates (like equality).
        return window.moment(chrono.parseDate(input)).startOf('day');
    }

    private stringIncludesCaseInsensitive(
        haystack: string,
        needle: string,
    ): boolean {
        return haystack
            .toLocaleLowerCase()
            .includes(needle.toLocaleLowerCase());
    }
}
