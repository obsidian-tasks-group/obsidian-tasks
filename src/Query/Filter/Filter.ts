import type { Task } from '../../Task';

/**
 * A filtering function, that takes a Task object and returns
 * whether it matches a particular filtering instruction.
 */
export type Filter = (task: Task) => boolean;

/**
 * A class which stores one of:
 * - A Filter
 * - An error message
 *
 * This is really currently a convenience for returning date from
 * Field.createFilterOrErrorMessage() and derived classes.
 *
 * Later, it may gain helper functions for constructing parser error messages,
 * as currently these are created by some rather repetitious code, and also
 * there is scope for making these messages more informative (including the
 * problem line, and perhaps listing allowed options).
 */
export class FilterOrErrorMessage {
    filter: Filter | undefined;
    error: string | undefined;

    /**
     * Construct a FilterOrErrorMessage with the filter.
     * @param filter
     */
    public static fromFilter(filter: Filter): FilterOrErrorMessage {
        const result = new FilterOrErrorMessage();
        result.filter = filter;
        return result;
    }

    /**
     * Construct a FilterOrErrorMessage with the given error message.
     * @param errorMessage
     */
    public static fromError(errorMessage: string): FilterOrErrorMessage {
        const result = new FilterOrErrorMessage();
        result.error = errorMessage;
        return result;
    }
}
