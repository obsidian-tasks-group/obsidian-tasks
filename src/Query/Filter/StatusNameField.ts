import type { Task } from '../../Task';
import { TextField } from './TextField';

/**
 * A {@link Field} implementation for searching status.name
 */
export class StatusNameField extends TextField {
    constructor() {
        super();
    }

    public fieldName(): string {
        return 'status.name';
    }

    value(task: Task): string {
        return task.status.name;
    }

    supportsSorting(): boolean {
        return true;
    }

    public supportsGrouping(): boolean {
        return true;
    }
}
