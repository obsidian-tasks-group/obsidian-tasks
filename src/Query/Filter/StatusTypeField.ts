import type { Task } from '../../Task';
import type { GrouperFunction } from '../Grouper';
import { StatusType } from '../../StatusConfiguration';
import type { Comparator } from '../Sorter';
import { Explanation } from '../Explain/Explanation';
import { Field } from './Field';
import { Filter, FilterOrErrorMessage } from './Filter';
import type { FilterFunction } from './Filter';

/**
 * A ${@link Field} implementation for searching status.type
 */
export class StatusTypeField extends Field {
    // -----------------------------------------------------------------------------------------------------------------
    // Filtering
    // -----------------------------------------------------------------------------------------------------------------
    public canCreateFilterForLine(line: string): boolean {
        // Use a relaxed regexp, just checking field name and not the contents,
        // so that we can parse the line later and give meaningful errors if user uses invalid values.
        const relaxedRegExp = new RegExp(`^(?:${this.fieldNameSingularEscaped()})`);
        return Field.lineMatchesFilter(relaxedRegExp, line);
    }

    createFilterOrErrorMessage(line: string): FilterOrErrorMessage {
        const match = Field.getMatch(this.filterRegExp(), line);
        if (match === null) {
            // It's OK to get here, because canCreateFilterForLine() uses a more relaxed regexp.
            return this.helpMessage(line);
        }

        const [_, filterOperator, statusTypeAsString] = match;

        const statusTypeElement = StatusType[statusTypeAsString.toUpperCase() as keyof typeof StatusType];
        if (!statusTypeElement) {
            return this.helpMessage(line);
        }

        let filterFunction: FilterFunction;

        switch (filterOperator) {
            case 'is':
                filterFunction = (task: Task) => {
                    return task.status.type === statusTypeElement;
                };
                break;
            case 'is not':
                filterFunction = (task: Task) => {
                    return task.status.type !== statusTypeElement;
                };
                break;
            default:
                return this.helpMessage(line);
        }

        return FilterOrErrorMessage.fromFilter(new Filter(line, filterFunction, new Explanation(line)));
    }

    protected filterRegExp(): RegExp | null {
        return new RegExp(`^(?:${this.fieldNameSingularEscaped()}) (is|is not) ([^ ]+)$`);
    }

    private helpMessage(line: string): FilterOrErrorMessage {
        const allowedTypes = Object.values(StatusType)
            .filter((t) => t !== StatusType.EMPTY)
            .join(' ');

        const message = `Invalid ${this.fieldNameSingular()} instruction: '${line}'.
    Allowed options: 'is' and 'is not' (without quotes).
    Allowed values:  ${allowedTypes}
                     Note: values are case-insensitive,
                           so 'in_progress' works too, for example.
    Example:         ${this.fieldNameSingular()} is not NON_TASK`;
        return FilterOrErrorMessage.fromError(line, message);
    }

    public fieldName(): string {
        return 'status.type';
    }

    value(task: Task): string {
        return task.status.type;
    }

    // -----------------------------------------------------------------------------------------------------------------
    // Sorting
    // -----------------------------------------------------------------------------------------------------------------

    supportsSorting(): boolean {
        return true;
    }

    comparator(): Comparator {
        return (a: Task, b: Task) => {
            const keyA = StatusTypeField.groupName(a);
            const keyB = StatusTypeField.groupName(b);
            return keyA.localeCompare(keyB, undefined, { numeric: true });
        };
    }

    // -----------------------------------------------------------------------------------------------------------------
    // Grouping
    // -----------------------------------------------------------------------------------------------------------------

    public supportsGrouping(): boolean {
        return true;
    }

    public grouper(): GrouperFunction {
        return (task: Task) => {
            return [StatusTypeField.groupName(task)];
        };
    }

    private static groupName(task: Task) {
        let prefix: string;
        // Add a numeric prefix to sort in to a meaningful order for users
        switch (task.status.type) {
            case StatusType.IN_PROGRESS:
                prefix = '1';
                break;
            case StatusType.TODO:
                prefix = '2';
                break;
            case StatusType.DONE:
                prefix = '3';
                break;
            case StatusType.CANCELLED:
                prefix = '4';
                break;
            case StatusType.NON_TASK:
                prefix = '5';
                break;
            case StatusType.EMPTY:
                prefix = '6';
                break;
        }
        return prefix + ' ' + task.status.type;
    }
}
