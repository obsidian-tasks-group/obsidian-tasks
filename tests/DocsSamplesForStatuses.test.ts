import { verify } from 'approvals/lib/Providers/Jest/JestApprovals';

import { Status } from '../src/Status';
import type { Task } from '../src/Task';
import { StatusConfiguration, StatusType } from '../src/StatusConfiguration';
import type { FilterOrErrorMessage } from '../src/Query/Filter/Filter';
import * as FilterParser from '../src/Query/FilterParser';
import { Group } from '../src/Query/Group';
import { StatusNameField } from '../src/Query/Filter/StatusNameField';
import { StatusTypeField } from '../src/Query/Filter/StatusTypeField';
import type { StatusCollection, StatusCollectionEntry } from '../src/StatusCollection';
import * as Themes from '../src/Config/Themes';
import { StatusValidator } from '../src/StatusValidator';
import { TaskBuilder } from './TestingTools/TaskBuilder';
import { MarkdownTable, verifyMarkdownForDocs } from './TestingTools/VerifyMarkdownTable';

function getPrintableSymbol(symbol: string) {
    const result = symbol !== ' ' ? symbol : 'space';
    return '`' + result + '`';
}

function verifyStatusesAsMarkdownTable(statuses: Status[], showQueryInstructions: boolean) {
    let statusName = 'Status Name';
    let statusType = 'Status Type';
    if (showQueryInstructions) {
        statusName += '<br>`status.name includes...`<br>`sort by status.name`<br>`group by status.name`';
        statusType += '<br>`status.type is...`<br>`sort by status.type`<br>`group by status.type`';
    }
    const table = new MarkdownTable([
        'Status Symbol',
        'Next Status Symbol',
        statusName,
        statusType,
        'Needs Custom Styling',
    ]);

    for (const status of statuses) {
        const statusCharacter = getPrintableSymbol(status.symbol);
        const nextStatusCharacter = getPrintableSymbol(status.nextStatusSymbol);
        const type = getPrintableSymbol(status.type);
        const needsCustomStyling = status.symbol !== ' ' && status.symbol !== 'x' ? 'Yes' : 'No';
        table.addRow([statusCharacter, nextStatusCharacter, status.name, type, needsCustomStyling]);
    }
    table.verifyForDocs();
}

function verifyStatusesAsTasksList(statuses: Status[]) {
    let markdown = '';
    for (const status of statuses) {
        const statusCharacter = getPrintableSymbol(status.symbol);
        markdown += `- [${status.symbol}] #task ${statusCharacter} ${status.name}\n`;
    }
    verifyMarkdownForDocs(markdown);
}

function verifyStatusesAsTasksText(statuses: Status[]) {
    let markdown = '';
    for (const status of statuses) {
        const statusCharacter = getPrintableSymbol(status.symbol);
        markdown += `- [${status.symbol}] #task ${statusCharacter} ${status.name}\n`;
    }
    verify(markdown);
}

function constructStatuses(importedStatuses: StatusCollection) {
    const statuses: Status[] = [];
    importedStatuses.forEach((importedStatus) => {
        statuses.push(Status.createFromImportedValue(importedStatus));
    });
    return statuses;
}

describe('DefaultStatuses', () => {
    // These "test" write out a markdown representation of the default task statuses,
    // for embedding in the user docs.
    it('core-statuses', () => {
        verifyStatusesAsMarkdownTable([Status.makeTodo(), Status.makeDone()], true);
    });

    it('custom-statuses', () => {
        verifyStatusesAsMarkdownTable([Status.makeInProgress(), Status.makeCancelled()], true);
    });

    it('important-cycle', () => {
        const importantCycle: StatusCollection = [
            ['!', 'Important', 'D', 'TODO'],
            ['D', 'Doing - Important', 'X', 'IN_PROGRESS'],
            ['X', 'Done - Important', '!', 'DONE'],
        ];
        verifyStatusesAsMarkdownTable(constructStatuses(importantCycle), false);
    });

    it('todo-in_progress-done', () => {
        const importantCycle: StatusCollection = [
            [' ', 'Todo', '/', 'TODO'],
            ['/', 'In Progress', 'x', 'IN_PROGRESS'],
            ['x', 'Done', ' ', 'DONE'],
        ];
        verifyStatusesAsMarkdownTable(constructStatuses(importantCycle), false);
    });

    it('pro-con-cycle', () => {
        const importantCycle: StatusCollection = [
            ['P', 'Pro', 'C', 'NON_TASK'],
            ['C', 'Con', 'P', 'NON_TASK'],
        ];
        verifyStatusesAsMarkdownTable(constructStatuses(importantCycle), false);
    });

    it('toggle-does-nothing', () => {
        const importantCycle: StatusCollection = [
            ['b', 'Bookmark', 'b', 'NON_TASK'],
            ['E', 'Example', 'E', 'NON_TASK'],
            ['I', 'Information', 'I', 'NON_TASK'],
            ['P', 'Paraphrase', 'P', 'NON_TASK'],
            ['Q', 'Quote', 'Q', 'NON_TASK'],
        ];
        verifyStatusesAsMarkdownTable(constructStatuses(importantCycle), false);
    });
});

describe('Theme', () => {
    type NamedTheme = [string, StatusCollection];
    const themes: NamedTheme[] = [
        // Alphabetical order by name:
        ['AnuPpuccin', Themes.anuppuccinSupportedStatuses()],
        ['Aura', Themes.auraSupportedStatuses()],
        ['Ebullientworks', Themes.ebullientworksSupportedStatuses()],
        ['ITS', Themes.itsSupportedStatuses()],
        ['Minimal', Themes.minimalSupportedStatuses()],
        ['Things', Themes.thingsSupportedStatuses()],
    ];

    describe.each(themes)('%s', (_: string, statuses: StatusCollection) => {
        it.each(statuses)('Validate status: "%s", "%s", "%s", "%s"', (symbol, name, nextSymbol, type) => {
            const statusValidator = new StatusValidator();
            const entry: StatusCollectionEntry = [symbol, name, nextSymbol, type];
            expect(statusValidator.validateStatusCollectionEntry(entry)).toEqual([]);
        });

        it('Table', () => {
            verifyStatusesAsMarkdownTable(constructStatuses(statuses), true);
        });

        it('Tasks', () => {
            verifyStatusesAsTasksList(constructStatuses(statuses));
        });

        it('Text', () => {
            verifyStatusesAsTasksText(constructStatuses(statuses));
        });
    });
});

function verifyTransitionsAsMarkdownTable(statuses: Status[]) {
    const columnNames: string[] = ['Operation and status.type'];
    statuses.forEach((s) => {
        const title = s.type;
        columnNames.push(title);
    });

    const table = new MarkdownTable(columnNames);

    const tasks: Task[] = [];
    {
        const cells: string[] = ['Example Task'];
        statuses.forEach((s) => {
            const task = new TaskBuilder().status(s).description('demo').build();
            tasks.push(task);
            cells.push('`' + task!.toFileLineString() + '`');
        });
        table.addRow(cells);
    }

    function filterAllStatuses(filter: FilterOrErrorMessage) {
        const cells: string[] = [`Matches \`${filter!.instruction}\``];
        tasks.forEach((task) => {
            const matchedText = filter!.filter?.filterFunction(task) ? 'YES' : 'no';
            cells.push(matchedText);
        });
        table.addRow(cells);
    }

    filterAllStatuses(FilterParser.parseFilter('not done')!);
    filterAllStatuses(FilterParser.parseFilter('done')!);
    filterAllStatuses(FilterParser.parseFilter('status.type is TODO')!);
    filterAllStatuses(FilterParser.parseFilter('status.type is IN_PROGRESS')!);
    filterAllStatuses(FilterParser.parseFilter('status.type is DONE')!);
    filterAllStatuses(FilterParser.parseFilter('status.type is CANCELLED')!);
    filterAllStatuses(FilterParser.parseFilter('status.type is NON_TASK')!);
    filterAllStatuses(FilterParser.parseFilter('status.name includes todo')!);
    filterAllStatuses(FilterParser.parseFilter('status.name includes in progress')!);
    filterAllStatuses(FilterParser.parseFilter('status.name includes done')!);
    filterAllStatuses(FilterParser.parseFilter('status.name includes cancelled')!);

    function showGroupNamesForAllTasks(groupName: string, grouperFunction: (task: Task) => string[]) {
        const cells: string[] = ['Name for `group by ' + groupName + '`'];
        tasks.forEach((task) => {
            const groupNamesForTask = grouperFunction(task);
            const names = groupNamesForTask.join(',');
            cells.push(names);
        });
        table.addRow(cells);
    }

    showGroupNamesForAllTasks('status', Group.grouperForProperty('status'));
    showGroupNamesForAllTasks('status.type', new StatusTypeField().createGrouper().grouper);
    showGroupNamesForAllTasks('status.name', new StatusNameField().createGrouper().grouper);

    table.verifyForDocs();
}

describe('Status Transitions', () => {
    it('status-types', () => {
        const statuses = [
            Status.makeTodo(),
            Status.makeInProgress(),
            Status.makeDone(),
            Status.makeCancelled(),
            new Status(new StatusConfiguration('~', 'My custom status', ' ', false, StatusType.NON_TASK)),
        ];
        verifyTransitionsAsMarkdownTable(statuses);
    });
});
