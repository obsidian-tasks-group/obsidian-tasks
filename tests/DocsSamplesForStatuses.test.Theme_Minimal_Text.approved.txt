- [ ] #task `space` to-do
- [/] #task `/` incomplete
- [x] #task `x` done
- [-] #task `-` canceled
- [>] #task `>` forwarded
- [<] #task `<` scheduling
- [?] #task `?` question
- [!] #task `!` important
- [*] #task `*` star
- ["] #task `"` quote
- [l] #task `l` location
- [b] #task `b` bookmark
- [i] #task `i` information
- [S] #task `S` savings
- [I] #task `I` idea
- [p] #task `p` pros
- [c] #task `c` cons
- [f] #task `f` fire
- [k] #task `k` key
- [w] #task `w` win
- [u] #task `u` up
- [d] #task `d` down
